### ISSUE TITLE
API rate restrictions when sending multiple emails to multiple recipients

### ISSUE THREAD

**Issue by carloshmm** [sendgrid/sendgrid-nodejs#1081]
### Issue Summary
The answer from this endpoint: https://sendgrid.api-docs.io/v3.0/mail-send/v3-mail-send has no **X-RateLimit** headers as it is said here: https://sendgrid.com/docs/API_Reference/api_v3.html

> Depending on the endpoint you are trying to reach, it will have a specific number of allowed requests per refresh period.

So I believe this specific endpoint does not have restrictions for requests per refresh period.

So I found this: https://sendgrid.api-docs.io/v3.0/mail-send/mail-send-limitations

> <ul>
> <li>The total size of your email, including attachments, must be less than 30MB.</li>
> <li>The total number of recipients must no more than 1000. This includes all recipients defined within the <code>to</code>, <code>cc</code>, and <code>bcc</code> parameters, across each object that you include in the <code>personalizations</code> array.</li>
> <li>The total length of custom arguments must be less than 10000 bytes.</li>
> <li>Unicode encoding is not supported for the <code>from</code> field.</li>
> <li>The <code>to.name</code>, <code>cc.name</code>, and <code>bcc.name</code> personalizations cannot include either the <code>;</code> or <code>,</code> characters.</li>
> </ul>

Investigating the code I found this code block: https://github.com/sendgrid/sendgrid-nodejs/blob/bce169432f7664968d8680c46c721fc515eccdca/packages/mail/src/classes/mail-service.js#L149-L166

It means that internally sendgrid-nodejs is just calling the endpoint v3/mail/send multiple times in parallel as it is said here: https://github.com/sendgrid/sendgrid-nodejs/blob/master/docs/use-cases/multiple-emails-multiple-recipients.md

> This will send multiple requests (in parallel), so be aware of any API rate restrictions

I'm configuring a back-end that will likely send up to 100.000 personalized emails following this strategy of sending multiple emails to multiple recipients. Since the endpoint v3/mail/send does not return any **X-RateLimit** headers and since I'm running this endpoint one time for each recipient I'm not hitting the restriction of **The total number of recipients must no more than 1000**.

Should I be afraid of any rate limiting?

**Comment by demian85** [comment]
I'd love to know anything related to this weird API. Documentation is a joke. I have so many questions that I don't know where to start.
I'm trying to do the same as you and I cannot figure out if there is any restriction limit when sending multiple emails and even worse, how do I track possible errors for each email? Is the promise rejected when an email fails or when all the emails fail? Are all emails sent in a single request? Why do I get a huge array of array of things as a response? I just want to know what the fuck succeeded or failed.

**Comment by childish-sambino** [comment]
More info here regarding mail send rate-limiting: https://sendgrid.com/docs/for-developers/sending-email/v3-mail-send-faq/#are-there-limits-on-how-often-i-can-send-email-and-how-many-recipients-i-can-send-to

> Currently, you may make up to 10,000 requests per second to our endpoint. Each email you send may include up to 1000 recipients.

Are they multiple emails you're looking at sending or could they be templatized with different recipients getting different dynamic data?

Regarding sending emails in parallel, `Promise.all` is not the right implementation if consumers are wanting to see details about each failure, but I think that warrants its own issue.

**Comment by carloshmm** [comment]
> Are they multiple emails you're looking at sending or could they be templatized with different recipients getting different dynamic data?
I'm building a system that will send templatized e-mails to about 10.000~100.000 recipients at once, the template system is built on the system itself, I'm not using the Sendgrid's template feature.
So If, for example, my system triggers an internal event of sending 100.000 different e-mails (templatized/personalized to each recipient) to 100.000 different recipients, should I restrict my system to send it in lots of 10.000 each 1 second?
> Regarding sending emails in parallel, `Promise.all` is not the right implementation if consumers are wanting to see details about each failure, but I think that warrants its own issue.
I don't understand what could be a failure, I'm currently validating the emails in advance, so I won't get any invalid e-mail errors and I'm sure I'm following the docs, is there something that I can't see?
Regarding e-mail feedbacks, I'm currently using the Sendgrid's Webhook event API, so I'll get errors if, for example, the inbox of one of the recipients is full.

I'd like to suggest using the new [Promise.allSettled](https://v8.dev/features/promise-combinators) method in place of Promise.all , so that the execution of the code "does not short-circuit" when something goes wrong, or at least create new new parameter for the send method that will allow this flow, is it a good idea?

**Comment by childish-sambino** [comment]
> should I restrict my system to send it in lots of 10.000 each 1 second?

Yes.

> I don't understand what could be a failure

Always plan for failure. Bugs in your code, bugs in ours, network hiccups, etc.

> I'd like to suggest using the new Promise.allSettled method in place of Promise.all

`Promise.allSettled` would be better, except it's no available until [Node v12.9.0](https://nodejs.org/en/blog/release/v12.9.0/). Probably okay to use a shim: https://www.npmjs.com/package/promise.allsettled

I'm open to reviewing a PR for this.

**Comment by jehunter5811** [comment]
I'm having an issue that follows much of the detail being discussed in this thread. We have built a service that allows customers to send emails to n number of their users. We have batched any requests with more than 1000 emails into groups of 1000 and use the sg.sengMultiple call on the NodeJS SDK. However, we continue to get the following error: 

Forbidden (403)

  You have exceeded your messaging limits

If each batch is 1000 emails (confirmed) and we have upgraded our plan to the 40,000/month plan, what else can we look for to help understand this problem?

**Comment by childish-sambino** [comment]
@jehunter5811 The `sendMultiple` implementation is not suited for this since it attempts to send all requests in parallel which could then exceed the rate limits. If you're needing to send large batches that could bump up against the limits, best to run them serially with delay between requests. If you're _still_ seeing rate limiting issues after that, best to contact [customer support](https://support.sendgrid.com).

**Comment by jehunter5811** [comment]
Interesting. Thanks for that. If I have 4 batches of 1000 emails, would that be 4 requests against the rate limiting or would it be 4000 requests? I've found docs on rate limits but not much in terms of what's considered a request.

**Comment by carloshmm** [comment]
> I'm open to reviewing a PR for this.

I'll think about it, maybe I'll try this solution and if it is so I'll open a PR.

I think I'll close this issue because my doubts were answered, if there is need to open it again just ask.

Thank you.

**Comment by childish-sambino** [comment]
@jehunter5811 A _request_ being an API POST call, which is a single call to `send` with a non-array payload.

**Comment by jehunter5811** [comment]
@childish-sambino got it! Thanks again and sorry for hijacking this thread. Super helpful though.
